### PR TITLE
Generalize ExportContext and use it for deprecation and unavailability

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/Sema/TypeCheckAvailability.h [hunk 1]
```diff
@@ -68,6 +69,29 @@ enum class ExportabilityReason : unsigned {
...
+/// declaration (such as function parameter and return types), but not
+/// including the function body.
+///
+/// The signature of an exported declaration can only reference other
+/// exported types.
+///
+/// The body of an inlinable function can only reference other `public` and
+/// `@usableFromInline` declarations; furthermore, if the inlinable
+/// function is also exported, its body is restricted to referencing other
+/// exported declarations.
```
Comment (xymus) [inline lib/Sema/TypeCheckAvailability.h:85]: When you say *if the inlinable function is also exported*, that would be for all inlinable functions except `@_spi` right?

File: lib/Sema/TypeCheckAvailability.h [hunk 2]
```diff
@@ -68,6 +69,29 @@ enum class ExportabilityReason : unsigned {
   ExtensionWithConditionalConformances
 };
 
+/// A description of the restrictions on what declarations can be referenced
+/// from a the signature or body of a declaration.
+///
+/// We say a declaration is "exported" if it is `public` or
+/// `@usableFromInline`, not `_@spi`, and not visible via an
```
Comment (xymus) [inline lib/Sema/TypeCheckAvailability.h:72]: There are typos here `_@spi` and above `from a the signature `.

**Review by xymus — APPROVED** [review]
LGTM and thank you for this work! It looks like it will be much easier to work with.

File: lib/Sema/TypeCheckAvailability.h (hunk 1 above)
Comment (slavapestov) [reply lib/Sema/TypeCheckAvailability.h:85]: Yep

File: lib/Sema/TypeCheckAvailability.h (hunk 2 above)
Comment (slavapestov) [reply lib/Sema/TypeCheckAvailability.h:72]: Thanks!

**Comment by slavapestov** [comment]
@xymus I'll fix the comments in my next PR since it's almost ready to go up anyway. Thanks for the review!

**Pull request merged**